Add mc_world_size

eruizc-dev opened the pull request:
I requested this feature a couple days ago, wasn't hard to do.

Closes #128

eruizc-dev commented:
Really cool graphs you can do with this
![image](https://user-images.githubusercontent.com/58179604/147162585-410a9607-e84e-4ed5-850b-fcddf000e103.png)

sladkoff commented on src/main/java/de/sldk/mc/metrics/WorldSize.java:
```diff
@@ -0,0 +1,43 @@
+
+    private long getFolderSize(File folder) {
+        long size = 0;
+        for (File file : folder.listFiles()) {
```
My IDE tells me that `folder.listFiles()` might produce a NullPointerException. Can you add a null check or consider using [this](https://stackoverflow.com/a/19877372/1510659) which seems more robust (but also more verbose :sweat_smile:)?

eruizc-dev replied:
I like that implementation, seem faster

sladkoff reviewed (changes requested):
Would it be possible for you to add a test or two? Thanks.

eruizc-dev commented:
Changed the implementation with that one, added way more tests than I anticipated because I wasn't exactly sure how that code worked and wanted to make sure it does not break.

I wasn't able to make unit tests for world size metric because I couldn't get around mocking PathUtils, but I guess it qualifies as integration test so I'm good with that.

eruizc-dev commented on src/main/java/de/sldk/mc/metrics/WorldSize.java:
```diff
@@ -0,0 +1,36 @@
+            long size = pathUtils.getSize();
+            String worldName = world.getName();
+            WORLD_SIZE.labels(worldName).set(size);
+        } catch (Throwable t) {
+            // ignore
+        }
```
Should this error be logged?

```
EROR: couldn't get world size metrics for world 'worl'
```

eruizc-dev replied:
Yeah, throwing does log at finer level. Done!

sladkoff commented on src/main/java/de/sldk/mc/utils/PathUtils.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.concurrent.atomic.AtomicLong;
+
+public class PathUtils {
+    private Path path;
```
Can be `final`.

sladkoff commented on src/test/java/de/sldk/mc/metrics/WorldSizeTest.java:
```diff
@@ -0,0 +1,111 @@
+    private World world;
+
+    @BeforeEach
+    public void beforeEach() throws IOException{
```
`throws IOException` can be removed.

sladkoff commented on src/test/java/de/sldk/mc/metrics/WorldSizeTest.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    @AfterEach
+    public void afterEach() {
+        CollectorRegistry.defaultRegistry.clear();
+        try {
+            Files.deleteIfExists(path);
+        } catch (Throwable t) { }
```
Here, I'd `throw` instead of an empty catch block. The test would fail in case of an Exception but I think that's ok in that case, no?

```suggestion
    public void afterEach() throws IOException {
        CollectorRegistry.defaultRegistry.clear();
        Files.deleteIfExists(path);
```

sladkoff commented on src/test/java/de/sldk/mc/metrics/WorldSizeTest.java:
```diff
@@ -0,0 +1,111 @@
+        assertNull(CollectorRegistry.defaultRegistry.getSampleValue(METRIC_NAME));
+    }
+
+    private void givenWorldFileDoesNotExist() throws IOException {
```
The `IOException` is never thrown here.

```suggestion
    private void givenWorldFileDoesNotExist() {
```

sladkoff commented on src/test/java/de/sldk/mc/utils/PathUtilsTest.java:
```diff
@@ -0,0 +1,131 @@
+        if (path != null) {
+            try {
+                Files.deleteIfExists(path);
+            } catch (IOException e) { }
```
Same here with the catch block. Just throw it?

sladkoff commented on src/test/java/de/sldk/mc/utils/PathUtilsTest.java:
```diff
@@ -0,0 +1,131 @@
+            assertEquals(0, PathUtils.getSize());
+        }
+
+        @Test
```
My IDE reports: `Suspicious combination @Test and @RepeatedTest`

sladkoff commented on src/test/java/de/sldk/mc/utils/PathUtilsTest.java:
```diff
@@ -0,0 +1,131 @@
+
+    private static File createFileInTmpDirectory(Path path, String name, int length) throws IOException {
+        File file = path.resolve(name).toFile();
+        FileWriter fileWriter = new FileWriter(file);
+        fileWriter.write(new RandomString(length).nextString());
+        fileWriter.close();
```
I'd make use of try-with-resources:

```suggestion
        try (FileWriter fileWriter = new FileWriter(file)) {
            fileWriter.write(new RandomString(length).nextString());
	}
```

sladkoff commented on src/test/java/de/sldk/mc/metrics/WorldSizeTest.java:
```diff
@@ -0,0 +1,111 @@
+        givenWorldFileExists(worldName);
+        String filename = new RandomString(10).nextString();
+        File f = path.resolve(filename).toFile();
+        FileWriter writer = new FileWriter(f);
+        writer.write(new RandomString(worldSize).nextString());
+        writer.close();
```
try-with-resources block could also be used here.

sladkoff reviewed (changes requested):
Looking good! I added some feedback. 

Could you also run a formatter? I've noticed missing spaces here and there :sweat_smile:

eruizc-dev commented:
I don't see where the spaces are missing, probably our formaters have different configurations. I've allowed maintainer modification, you can run yours no problem!

sladkoff commented on src/main/java/de/sldk/mc/utils/PathFileSize.java:
```diff
@@ -0,0 +1,32 @@
+
+    public long getSize() throws IOException {
+        final AtomicLong size = new AtomicLong(0);
+        Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                size.addAndGet(attrs.size());
+                return FileVisitResult.CONTINUE;
+            }
+
+            public FileVisitResult visitFileFailed(Path file, IOException exc) {
+                return FileVisitResult.CONTINUE;
+            }
+        });
```
Can you check if you could refactor the temp file cleanups with the JUnit feature `@TempDir`. See [this StackOverflow answer](https://stackoverflow.com/a/56614039/1510659). I think it would become much simpler.

eruizc-dev replied:
I've been looking for that and could not find it, thanks for finding it

sladkoff commented:
@eruizc-dev It's perfect :star: 

I'm so sorry for the delay.

sladkoff commented:
Released in [2.5.0](https://github.com/sladkoff/minecraft-prometheus-exporter/releases/tag/v2.5.0) :tada:
